### PR TITLE
Zoombutton

### PR DESCRIPTION
**Pull request by Tiomat85**
Initial draft pull request to run tests.

### CODE REVIEW

**Comment by cmeyer** [comment]
Can you use these zoom in/out icons instead (you may have to modify size to be 24x24 to conform to other tool icons):

zoom_in_cursor
zoom_out_cursor

**Review by cmeyer — CHANGES_REQUESTED** [review]
I know this is still draft, and I'm only reviewing the cursor icons.

**Comment by cmeyer** [comment]
I didn't get to this today - but a question anyway: is this ready for me to review?

**Comment by Tiomat85** [comment]
> I didn't get to this today - but a question anyway: is this ready for me to review?

Yes it should be ready for a first review, I am still trying to figure if I can add the drag-to-zoom stuff but that will come in as a second PR.

File: nion/swift/ImageCanvasItem.py [hunk 1]
```diff
@@ -639,6 +639,70 @@ async def _reactor_loop(self, r: Stream.ValueChangeStreamReactorInterface[MouseP
                 change_display_properties_task.commit()
 
 
+class ZoomMouseHandler(MouseHandler):
+    def __init__(self, image_canvas_item: ImageCanvasItem, event_loop: asyncio.AbstractEventLoop, is_zooming_in: bool) -> None:
+        super().__init__(image_canvas_item, event_loop)
+        self.cursor_shape = "mag_glass"
```
[the remark this answers is not in the document]
Comment (cmeyer) [reply nion/swift/ImageCanvasItem.py:645]: Almost... can you remove the `self.cursor_shape = "mag_glass"` line above?

File: nion/swift/test/ImageCanvasItem_test.py [hunk 2]
```diff
@@ -291,6 +291,27 @@ def test_hand_tool_on_one_image_of_multiple_displays(self):
             document_controller.tool_mode = "hand"
             display_panel.display_canvas_item.simulate_press((100,125))
 
+    def test_zoom_tool_on_one_image_of_multiple_displays(self):
```
Comment (cmeyer) [inline nion/swift/test/ImageCanvasItem_test.py:294]: Can you rename this to something like `test_zoom_in_and_out_around_clicked_point`? The `one_image_of_multiple_displays` is spurious.

**Review by cmeyer — CHANGES_REQUESTED** [review]
After trying this branch, a few comments:

I think we should put a single magnifying glass icon in the tool bar: zoom in by 25% by default; out by 25% if the shift key is held down. It shouldn't have a "plus" or "minus" on the inside.

Should we also use the 'z' key to select the zoom tool?

When clicking with the zoom tool, it doesn't zoom in/out around the point that you click. You'll need to adjust the `image_position` value such that `ImageCanvasItemMapping.map_point_widget_to_image` for the mouse point is invariant before and after the zoom. I didn't look closely at the algorithm, but I think you can track the necessary information down from this line:

https://github.com/nion-software/nionswift/blob/c142d861e12f7743ad304b190109e710285181cf/nion/swift/ImageCanvasItem.py#L978

Clicking multiple times should keep the click-upon feature exactly under the same spot under the cursor.

Adding some checks to the test for this would be good.

File: nion/swift/ImageCanvasItem.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tiomat85) [reply nion/swift/ImageCanvasItem.py:645]: Fixed, not sure how I missed that.

File: nion/swift/test/ImageCanvasItem_test.py (hunk 2 above)
Comment (Tiomat85) [reply nion/swift/test/ImageCanvasItem_test.py]: Test has been renamed, and the mag glass reference removed.
The zooming maintaining the cursor position has been sorted out, that function that you mentioned was a bit of a red herring but I could jiggle around my mapping of widget to image coordinate spaces anyway to get what was needed for the extra vector needing to be applied.

On the single button option, that isn't quite what I thought was agreed as things seem to reference buttons plural. Changing it to a single zoom button would make it pretty opaque on how to use it (needing to press random buttons to find out how to zoom out). It also begs the question about the two icons you sent me to use :)  For the time being it would make sense to go with the two button option, accept that we know it isn't perfect but it is still better than not having buttons at all. With the way the discussions were going anyway, the buttons are likely to be irrelevant entirely and simply be replaced with using the scroll-wheel on the mouse and the buttons being a backup.

On the 'z' shortcut key, personally I am not a fan of shortcut keys that can be accidently activated as it can lead to confusion. I know you have some existing shortcut keys that are single key ones and if that is your preferred pattern then for the future block of work we could add them in but again it may become redundant with the scroll wheel.

Comment (cmeyer) [reply nion/swift/test/ImageCanvasItem_test.py]: Thanks for renaming the test. Can you add code to the test to ensure that it is actually zooming around the clicked point? I think there are similar examples in the `ImageCanvasItem_test.py` file, but if you need specific references, let me know.

Regarding the single button for the tool: zooming in/out on the same point is often required to get the view just right; having two tools requires moving the mouse back to the tool bar to switch, losing the zoom point in the process. As far as I can see from a quick search, the industry norm is a single tool in software that uses tools. Some software allows you to place a point and zoom in/out on that point with a keybaord. We can consider that in the future, i.e. "zoom around graphic". I couldn't find another program that has two zoom tools. The option and control keys already modify many of the other tools, and it's an industry standard, so I don't think it's a big stretch for users to discover the functionality. We can add a cursor that changes in the future to make it easier. I'd also like double clicking the zoom tool to reset the display to 'fit'. And eventually dragging to fill the display panel with the dragged area.

I personally don't like scroll-wheel-zooming, but we can try it out in another PR. However, I think we should spend our time on this PR to get it right, without assuming it will be replaced by something else in the future.

Regarding the shortcut keys - I'm open to different ideas. I'm not a fan of the single keys affecting modes either. Maybe we do something like multiple keys? Esc+'z' to switch modes? We definitely will need some sort of key shortcut for this and other tools. Let's open a discussion offline.

The `mag_glass_in`/`mag_glass_out` files are spurious. And I've attached a replacement for the magnifying glass.

zoom
zoom

Comment (Tiomat85) [reply nion/swift/test/ImageCanvasItem_test.py]: I can certainly look to add extra checks to the test, though I have a query as to what these tests are trying to actually verify.  I mirrored the 'hand tool' test, almost exactly so I am trying to understand what these test are attempting to validate. The 'hand tool' test as far as I can see validates that the code path is runnable (as do other tests in the ImageCanvasItem_test file), with no defined asserts so I was working on the idea that the tests validate all code paths are runnable, with extra asserts added to validate state easily.  If instead the objective is to try and go full TDD, then surely we need the test requirements pre-defined as part of the Issue definition which can then be used as verification of completion of functionality.

Shortcut-keys, they tend to be either Ctrl-X or Alt-X as holding that extra key is something explicit the user has to do and is unlikely to be caught by accident.  My concern with just the single keys is it is quite easy to knock one by accident and suddenly whichever image you had selected may/may not have changed in some unforeseen way. If it is consistent with the way Swift works then we can add it in, just trying to work through the 'question everything' phase to understand why.

Comment (Tiomat85) [reply nion/swift/test/ImageCanvasItem_test.py]: On the single Zoom icon, I can certainly change it to a single icon instead and we can go off left-click/right-click as the convention on quite a bit of software seems to be. This especially seems to be the case in more modern software to go with a single icon rather than two, in parallel with the scroll-to-zoom which now seems to be everywhere.  My reasoning for pushing back is the original discussion was always about buttons plural, and you yourself sent me two icons, so now with it suddenly switching to wanting a single icon I am trying to understand the reasoning for the change both from a 'are we building the right thing' but also a process point of view. If it is simply that the original requirements were wrong, and you hadn't thought about the implications of the icons you were sending, then that is ok but trying to make sure we stay on the same page here.

Since scroll-to-zoom wasn't included in the original Issue as a concrete point, just some discussion I am going to park those for the moment until there is consensus on it. Similarly with the 'zoom around graphic', double-click to fit, drag-to-zoom (though code-behind for this is already done, minus the drawing of the selection rectangle).  If we want to go whole-hog on this Pull Request and fix everything to do with zooming with it, then I am ok with that but we should recognise that we are somewhat deviating from the process of fixing a well defined Issue and going with a more fluid 'experimental' UX development.

Comment (cmeyer) [reply nion/swift/test/ImageCanvasItem_test.py]: I'm not proposing to expand this PR in any way - just to implement it in a reasonable way. Zoom tool that zooms in/out with adequate test. No key overload (for now). No dragging (for now). No zoom around graphic (for now). No zoom-specific cursor (for now).

The specific test to copy is this one: `test_hand_tool_undo` in `DisplayPanel_test.py`. Please put it in `DisplayPanel_test.py`, even though that might not be the ideal spot. Its good to keep the two tests together.

Regarding the tool: here's the gold standard for the zoom tool in applications like this:
https://helpx.adobe.com/illustrator/using/tool-techniques/zoom-tool.html

I've never seen a program use click/right-click in a zoom button. But either way, right-click is overloaded to do the context menu in Swift. Also, track pads (laptops) click/right-click would be burdensome. Let's just follow the gold standard linked above.

Originally I provided two icons because I didn't think through the issue. After trying it out, it is really obvious that we should have a single zoom tool that switches when holding the Alt/Opt key.

**Comment by cmeyer** [comment]
Can we move this PR forward or close it? Mainly this should be a single 'zoom' tool that zooms in when clicked and zooms out when alt-clicked (Windows). The code looks like it has an `if False...` in it, too, that should be removed if unused.

**Comment by cmeyer** [comment]
I added #1137 to limit the scope of this change and added it to the 16.12 milestone. This PR is very close, just needs to replace the two tools with the one and zoom out on Alt+click.

**Comment by Tiomat85** [comment]
Aligned the PR along with the discussed scope.

File: nion/swift/ImageCanvasItem.py [hunk 3]
```diff
@@ -1157,6 +1257,7 @@ def mouse_position_changed(self, x: int, y: int, modifiers: UserInterface.Keyboa
         image_position = widget_mapping.map_point_widget_to_image(mouse_pos)
         if delegate.image_mouse_position_changed(image_position, modifiers):
             return True
+
```
Comment (KRLango) [inline nion/swift/ImageCanvasItem.py:1260]: Is there a reason to add a blank line here?

File: nion/swift/resources/mag_glass_in.png
Comment (KRLango) [inline nion/swift/resources/mag_glass_in.png:1]: Do we want all 3 of these images? I thought we were just using 1.

File: nion/swift/test/ImageCanvasItem_test.py [hunk 4]
```diff
@@ -291,6 +292,27 @@ def test_hand_tool_on_one_image_of_multiple_displays(self):
...
+            header_height = display_panel.header_canvas_item.header_height
+            display_panel.root_container.layout_immediate((1000 + header_height, 1000))
+            # run test
+            document_controller.tool_mode = "zoom"
+            display_panel.display_canvas_item.simulate_press((100, 125))
+            display_panel.display_canvas_item.simulate_release((100, 125))
+
+            document_controller.tool_mode = "zoom"
+            display_panel.display_canvas_item.simulate_press((125, 100), KeyboardModifiers.control)
+            display_panel.display_canvas_item.simulate_release((125, 100), KeyboardModifiers.control)
```
Comment (KRLango) [inline nion/swift/test/ImageCanvasItem_test.py:315]: This tests the click doesn't cause an exception. Is there a property we can check to ensure that the zoom has happened?

**Review by KRLango — CHANGES_REQUESTED** [review]

File: nion/swift/resources/mag_glass_in.png
Comment (Tiomat85) [reply nion/swift/resources/mag_glass_in.png:1]: We are, I missed the pngs.

File: nion/swift/ImageCanvasItem.py (hunk 3 above)
Comment (Tiomat85) [reply nion/swift/ImageCanvasItem.py:1260]: Preference for white space, but other than that no. Removed.

**Comment by Tiomat85** [comment]
So there is one known issue currently where if you zoom out (using Ctrl) on a different display item to the one currently selected, sometimes it adds that item to the current selection.  Trying to avoid this using other modifiers conflicted with other functionality so shortcuts as a whole may need investigating at some point to try and remove the complexity.

File: nion/swift/ImageCanvasItem.py [hunk 5]
```diff
@@ -639,6 +639,43 @@ async def _reactor_loop(self, r: Stream.ValueChangeStreamReactorInterface[MouseP
...
+        # get the beginning mouse position
+        value_change = await r.next_value_change()
+        value_change_value = value_change.value
+        assert value_change.is_begin
+        assert value_change_value is not None
+        image_position: typing.Optional[Geometry.FloatPoint] = None
+        mouse_pos, modifiers = value_change_value
+
+        if modifiers.control:
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:659]: This should be `modifiers.alt`, mapping to the Alt key on Windows and the Option key on macOS. See https://helpx.adobe.com/illustrator/using/tool-techniques/zoom-tool.html.

File: nion/swift/ImageCanvasItem.py [hunk 6]
```diff
@@ -1073,15 +1110,66 @@ def _update_image_canvas_position(self, widget_delta: Geometry.FloatSize) -> Geo
...
         return new_image_canvas_position
 
+    def convert_pixel_to_normalised(self, coord: Geometry.IntPoint) -> Geometry.FloatPoint:
+        if coord:
+            widget_mapping = ImageCanvasItemMapping.make(self.__data_shape, self.__composite_canvas_item.canvas_bounds,
+                                                         list())
+            if widget_mapping:
+                mapped = self.map_widget_to_image(coord)
+                if (mapped is not None) and (self.__data_shape is not None):
+                    norm_coord = tuple(ele1 / ele2 for ele1, ele2 in zip(iter(mapped), iter(self.__data_shape)))
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:1120]: Can you name these "y"  instead of "ele1" and "x"  instead of "ele2" ? I'm not sure what "ele" means but it is confusing to a casual code reviewer.

File: nion/swift/ImageCanvasItem.py [hunk 7]
```diff
@@ -1073,15 +1110,66 @@ def _update_image_canvas_position(self, widget_delta: Geometry.FloatSize) -> Geo
...
+                    return Geometry.FloatPoint(norm_coord[0], norm_coord[1])  # y,x
+        return Geometry.FloatPoint(-1, -1)
+
+    # Apply a zoom factor to the widget, optionally focussed on a specific point
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:1124]: "focussed" is misspelled. "focused"

File: nion/swift/ImageCanvasItem.py [hunk 8]
```diff
@@ -1073,15 +1110,66 @@ def _update_image_canvas_position(self, widget_delta: Geometry.FloatSize) -> Geo
...
+    def apply_fixed_zoom(self, zoom_in: bool, coord: typing.Optional[Geometry.IntPoint]) -> None:
+        zoom_factor = 0.25
+        if coord:
+            # Coordinate specified, so needing to recenter to that point before we adjust zoom levels
+            widget_mapping = ImageCanvasItemMapping.make(self.__data_shape, self.__composite_canvas_item.canvas_bounds, list())
+            if widget_mapping:
+                # coord is (x,y)
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:1131]: "coord is (x,y)" -> I don't think so - can you double check that?

File: nion/swift/ImageCanvasItem.py [hunk 9]
```diff
@@ -1073,15 +1110,66 @@ def _update_image_canvas_position(self, widget_delta: Geometry.FloatSize) -> Geo
...
+                        else:
+                            vector_scaling = zoom_factor
+
+                        scaled_mapped_vector = (mapped_vector[0] * vector_scaling, mapped_vector[1] * vector_scaling)
+                        if zoom_in:
+                            new_mapped_center = (mapped_center[0] - scaled_mapped_vector[0], mapped_center[1] - scaled_mapped_vector[1])
+                        else:
+                            new_mapped_center = (mapped_center[0] + scaled_mapped_vector[0], mapped_center[1] + scaled_mapped_vector[1])
+
+                        norm_coord = tuple(ele1 / ele2 for ele1, ele2 in zip(iter(new_mapped_center), iter(self.__data_shape)))
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:1148]: Same comment here: use y/x in place of ele1/ele2.

File: nion/swift/ImageCanvasItem.py [hunk 10]
```diff
@@ -1073,15 +1110,66 @@ def _update_image_canvas_position(self, widget_delta: Geometry.FloatSize) -> Geo
...
+                        self._set_image_canvas_position(Geometry.FloatPoint(norm_coord[0], norm_coord[1]))
+
+                        # ensure that at least half of the image is always visible
+                        new_image_norm_center_0 = max(min(norm_coord[0], 1.0), 0.0)
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:1152]: Likewise - how about "new_image_norm_center_y/x"?

File: nion/swift/ImageCanvasItem.py [hunk 11]
```diff
@@ -1127,7 +1221,13 @@ def mouse_released(self, x: int, y: int, modifiers: UserInterface.KeyboardModifi
         if self.__mouse_handler:
             self.__mouse_handler.mouse_released(Geometry.IntPoint(y, x), modifiers)
             self.__mouse_handler = None
-        if delegate.tool_mode != "hand":
+
+        # Should probably wrap this into a function of 'Non-Toggle' UI elements
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:1225]: This comment is not wrong, but I'd prefer to just make another PR with the change, or put it in an issue instead of making to-do items in the code that will never get done. I used to do the same thing, and almost none of them get done, so my preference now is to file an issue for "internal maintenance."

File: nion/swift/test/ImageCanvasItem_test.py [hunk 12]
```diff
@@ -291,6 +292,27 @@ def test_hand_tool_on_one_image_of_multiple_displays(self):
...
+            display_item = document_model.get_display_item_for_data_item(data_item)
+            copy_display_item = document_model.get_display_item_copy_new(display_item)
+            display_panel.set_display_panel_display_item(copy_display_item)
+            header_height = display_panel.header_canvas_item.header_height
+            display_panel.root_container.layout_immediate((1000 + header_height, 1000))
+            # run test
+            document_controller.tool_mode = "zoom"
+            display_panel.display_canvas_item.simulate_press((100, 125))
+            display_panel.display_canvas_item.simulate_release((100, 125))
+
```
Comment (cmeyer) [inline nion/swift/test/ImageCanvasItem_test.py:312]: What's missing from this test is confirmation that the zoom tool is performing as expected, i.e. by checking image position and image zoom. In addition, I'd like to see a test with the other view modes as initial conditions (i.e. 1:1 mode, etc.). A quick test of the functionality reveals to me some problems that should be caught by (a) manual testing and (b) automated tests here:

- zoom out gets way off center after about 4 zoom outs
- starting with 1:1 zooming (press '1' in panel to get to that state) or 'fill' zoom mode (press '0') do not give expected results.

See also this issue, which may need to go in before this PR:
- #1169

File: nion/swift/DocumentController.py [hunk 13]
```diff
@@ -3292,6 +3292,7 @@ def execute(self, context: Window.ActionContext) -> Window.ActionResult:
 Window.register_action(SetToolModeAction("wedge", _("Wedge"), "wedge_icon.png", _("Wedge tool for creating wedge masks")))
 Window.register_action(SetToolModeAction("ring", _("Ring"), "annular_ring.png", _("Ring tool for creating ring masks")))
 Window.register_action(SetToolModeAction("lattice", _("Lattice"), "lattice_icon.png", _("Lattice tool for creating periodic lattice masks")))
+Window.register_action(SetToolModeAction("zoom", _("Zoom"), "zoom.png", _("Zoom in/out on image")))
```
Comment (cmeyer) [inline nion/swift/DocumentController.py:3295]: I'm not sure of the perfect overall ordering since we now have an odd number of general tools (pointer, zoom, hand), and this might need rework in the future, but I think the best way is to put "zoom" as the 3rd item in the list, after "hand". That would put it at the top of the 2nd column in the tools. It should definitely not be the last item.

In addition, the tool window is no longer aligned when it has an odd number of items. I think this can be fixed easily with this patch, which should be included in this PR:

```
diff --git a/nion/swift/ToolbarPanel.py b/nion/swift/ToolbarPanel.py
index 1dd95410..a7fbf41d 100644
--- a/nion/swift/ToolbarPanel.py
+++ b/nion/swift/ToolbarPanel.py
@@ -67,6 +67,8 @@ class ToolModeToolbarWidget(Declarative.Handler):
                 top_row_items.append(radio_button)
             else:
                 bottom_row_items.append(radio_button)
+        if len(tool_actions) % 2 == 1:
+            bottom_row_items.append(u.create_spacing(32))
 
         top_row = u.create_row(*top_row_items)
         bottom_row = u.create_row(*bottom_row_items)
```

**Review by cmeyer — CHANGES_REQUESTED** [review]
Overall much closer to the final version! There are some details to fix, though, in this comment and others inline with the code.

Can you use the zoom icon I provided earlier? Here is the SVG and PNG. This is thicker, smaller so that it matches the size of the other tool icons, and hand editable without all of the inkscape cruft.

```
<?xml version="1.0"?>
<svg width="24" height="24"
     viewPort="0 0 24 24" version="1.1"
     xmlns="http://www.w3.org/2000/svg">

    <!-- Cursor -->
    <g transform="scale(0.8333,0.8333) translate(1, 4)">
    <circle cx="9" cy="9" r="6" stroke="black" fill="none" stroke-width="2"/>
    <path d="M 22,20 L 20,22 L 14,16 L 14,15 L 13,14 L 14,13 L 15,14 L 16,14 z" fill="black" stroke-linejoin="round"/>
    </g>
    <!-- <rect x="0" y="0" width="24" height="24" fill="none" stroke="green" stroke-width="0.5"/> -->

</svg>
```

zoom

File: nion/swift/ImageCanvasItem.py (hunk 6 above)
Comment (Tiomat85) [reply nion/swift/ImageCanvasItem.py:1279]: y and x in this context wouldn't be anymore descriptive (and potentially confusing) as they are not the X and Y parts.  I have renamed them to something hopefully a bit more descriptive (mapped_coord and shape_dim) to try represent the first one is the coordinate in mapped space, and the second one is the full dimension of the data to get a normalised value.

File: nion/swift/ImageCanvasItem.py (hunk 8 above)
Comment (Tiomat85) [reply nion/swift/ImageCanvasItem.py:1290]: You're right, was getting confused by the debugger ordering by name. First element is 'y'.  It is accessed by name anyway, so was only comment based incorrectness.

File: nion/swift/ImageCanvasItem.py (hunk 9 above)
Comment (Tiomat85) [reply nion/swift/ImageCanvasItem.py:1307]: See above, changed to new_mapped_coord and shape_dim.

**Review by cmeyer — APPROVED** [review]

File: nion/swift/ImageCanvasItem.py [hunk 14]
```diff
@@ -801,6 +801,43 @@ async def _reactor_loop(self, r: Stream.ValueChangeStreamReactorInterface[MouseP
...
+        value_change_value = value_change.value
+        assert value_change.is_begin
+        assert value_change_value is not None
+        image_position: typing.Optional[Geometry.FloatPoint] = None
+        mouse_pos, modifiers = value_change_value
+
+        if modifiers.alt:
+            is_zooming_in = False
+        else:
+            is_zooming_in = True
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:824]: `is_zooming_in = not modifiers.alt`, or better: `is_zooming_out = modifiers.alt`

File: nion/swift/ImageCanvasItem.py [hunk 15]
```diff
@@ -801,6 +801,43 @@ async def _reactor_loop(self, r: Stream.ValueChangeStreamReactorInterface[MouseP
...
+
+        with (delegate.create_change_display_properties_task() as change_display_properties_task):
```
Comment (cmeyer) [inline nion/swift/ImageCanvasItem.py:826]: Can you remove the parenthesis around the clause to make it more idiomatic Python?

File: nion/swift/test/ImageCanvasItem_test.py (hunk 4 above)
Comment (cmeyer) [inline nion/swift/test/ImageCanvasItem_test.py:315]: Unfortunately, this test is not testing anything nor even doing the zoom. As a quick explanation, the zoom tool is using an async reactor to handle the mouse tracking; but the test would need to call `document_controller.periodic()` after each mouse click to run the reactor. In addition, the keyboard modifiers values being passed `KeyboardModifiers.control` are actually pointers to abstract functions (not even calling the functions - just pointers to them).

Here is a test which actually performs the zoom and checks if it nominally works correctly (it fails unfortunately). We'll need several more tests like this to cover the other cases:

```
    def test_zoom_tool_in_and_out_around_clicked_point(self):
        with TestContext.create_memory_context() as test_context:
            document_controller = test_context.create_document_controller()
            document_model = document_controller.document_model
            display_panel = document_controller.selected_display_panel
            data_item = DataItem.DataItem(numpy.zeros((50, 50)))
            document_model.append_data_item(data_item)
            display_item = document_model.get_display_item_for_data_item(data_item)
            display_panel.set_display_panel_display_item(display_item)
            header_height = display_panel.header_canvas_item.header_height
            display_panel.root_container.layout_immediate((100 + header_height, 100))
            # run test. each data pixel will be spanned by two display pixels. clicking at an odd value
            # will be a click in the center of a data pixel. zoom should expand and contract around that
            # pixel. there is also a check to make sure zooming is not just ending up with the same exact
            # zoom, which would also succeed.
            self.assertEqual((25, 25), display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(51, 51)))
            self.assertEqual((30, 30), display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(61, 61)))
            document_controller.tool_mode = "zoom"
            display_panel.display_canvas_item.simulate_click((26, 26))
            # the zoom tool runs asynchronously, so give it a slice of async time.
            document_controller.periodic()
            # check results
            self.assertEqual((25, 25), display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(51, 51)))
            self.assertNotEqual((30, 30), display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(61, 61)))
            # zoom out is done by passing the alt key to the canvas item.
            display_panel.display_canvas_item.simulate_click((26, 26), CanvasItem.KeyboardModifiers(alt=True))
            document_controller.periodic()
            # zoom in at the center of a data pixel followed by zoom out in the same spot should end up in the same original mapping.
            self.assertEqual((25, 25), display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(51, 51)))
            self.assertEqual((30, 30), display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(61, 61)))
```

**Review by cmeyer — CHANGES_REQUESTED** [review]
Thanks for fixing the previous set of issues. I have a couple of minor coding comments; but mostly I've added a test which checks the actual zooming behavior from 'fit' mode with a centered image - and it shows that the zoom is not actually taking place around the cursor position. We'll need a bunch more tests - unfortunately zooming isn't as trivial to get right.

File: nion/swift/test/ImageCanvasItem_test.py (hunk 4 above)
Comment (cmeyer) [reply nion/swift/test/ImageCanvasItem_test.py:315]: After looking this further, I don't think the current layout system is going to be able to support proper zoom. We'll need something like the test above, but there will be limitations because the layout uses the canvas size for zooming, and canvas size is rounded to integer boundaries. Let's do the best we can with the current limitations, but zooming in/out multiple times is going to be off center due to rounding errors. We will fix that in a future PR by changing the layout and composition to accommodate fractional coordinates via scaling and translation.

File: nion/swift/ImageCanvasItem.py (hunk 14 above)
Comment (Tiomat85) [reply nion/swift/ImageCanvasItem.py:824]: It is now, it used to be more but has slowly shrunk to that and we can probably invert it to out like you say.

File: nion/swift/test/ImageCanvasItem_test.py (hunk 4 above)
Comment (Tiomat85) [reply nion/swift/test/ImageCanvasItem_test.py:315]: Indeed, you can see this if you zoom in really far (single digit displayed pixels on a full size canvas) you can really see that.  I think there are two distinct problems I identified:
Since it renders entire pixels at large zooms, even with the zoom factor being saved correctly number of pixels rendered doesn't always agree.
Again, when zoomed in a lot the 'coordinate' clicked becomes a more complex question. Was the click on the top left of a pixel, the center of the pixel, or does it need to have 'full resolution' of where abouts in the pixel did you click.

File: nion/swift/test/ImageCanvasItem_test.py [hunk 16]
```diff
@@ -291,6 +304,47 @@ def test_hand_tool_on_one_image_of_multiple_displays(self):
...
+            display_panel.display_canvas_item.simulate_click((26, 26))
+            # the zoom tool runs asynchronously, so give it a slice of async time.
+            document_controller.periodic()
+            # check results
+            self.assertTupleAlmostEqual((25, 25),
+                             display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(51, 51)),
+                             delta=math.ceil((25*0.1)+1))
+            self.assertTupleAlmostEqual((30, 30),
+                                display_panel.display_canvas_item.map_widget_to_image(Geometry.IntPoint(61, 61)),
+                                delta=math.ceil((30*0.1)+1))
```
Comment (cmeyer) [inline nion/swift/test/ImageCanvasItem_test.py:337]: Why did you make this section _less_ restrictive? In doing so, I think you've actually added a bug to the tests. If you zoom out around 51,51, then the mapping at 61,61 should NOT be equal (see my original test). But your test has changed it to test to be equal - which means your test is succeeded where it should be failing...?

Comment (cmeyer) [reply nion/swift/test/ImageCanvasItem_test.py:337]: Upon further inspection, `math.ceil((30*0.1)+1)` gives a range of +/- 4 (!) pixels. That's just wrong. It should be far more restrictive than that even if you don't want to check exact floating point numbers. Please tell me this wasn't an AI tool that suggested that code...? A reasonable range from be 10e-3 at first glance.

**Review by cmeyer — CHANGES_REQUESTED** [review]
Comments say it all.

File: nion/swift/ImageCanvasItem.py [hunk 17]
```diff
@@ -1232,15 +1265,67 @@ def _update_image_canvas_position(self, widget_delta: Geometry.FloatSize) -> Geo
...
+                        scaled_mapped_vector = (mapped_vector[0] * vector_scaling, mapped_vector[1] * vector_scaling)
+                        if zoom_in:
+                            new_mapped_center = (mapped_center[0] - scaled_mapped_vector[0], mapped_center[1] - scaled_mapped_vector[1])
+                        else:
+                            new_mapped_center = (mapped_center[0] + scaled_mapped_vector[0], mapped_center[1] + scaled_mapped_vector[1])
+
+                        norm_coord = tuple(new_mapped_coord / shape_dim for new_mapped_coord, shape_dim in zip(iter(new_mapped_center), iter(self.__data_shape)))
+                        self._set_image_canvas_position(Geometry.FloatPoint(norm_coord[0], norm_coord[1]))
+
+                        # ensure that at least half of the image is always visible
```
[the remark this answers is not in the document]
Comment (Tiomat85) [reply nion/swift/ImageCanvasItem.py:1307]: This is a carry over from the template function recommended, and I assumed it was a UI based requirements. Removed.

File: nion/swift/test/ImageCanvasItem_test.py (hunk 16 above)
Comment (Tiomat85) [reply nion/swift/test/ImageCanvasItem_test.py:337]: This was an error in the unit test case that it was not verifying against the wrong numbers causing the test to fail unexpectedly. This has been resolved, and we can discuss the remaining discrepancy.

**Comment by cmeyer** [comment]
Squashed and merged here c2a389a334136871d3bfd513a7e40a0f8dbd646c
